dotnet/src/SemanticKernel/AI/OpenAI/Clients/RetryHandler.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.SemanticKernel.AI.OpenAI.Clients;
+
+internal class RetryHandler: DelegatingHandler
+{
+    private const string Retry_After = "Retry-After";
+    private const int MaxRetries = 10;
```

Comment:
In the worst case, this handler will delay original call to a downstream service by 1.6 minute (10 retries * 10 seconds delay time) + 10 retries * each request latency. It's a lot and might be inacceptable for majority user facing/related  operations where the overall delay should not exceed a few seconds. It's a usual practice to fail fast by doing a few retry attempts with a 500ms-1s delay between them and let the exception bubble up the call chain up to the point where there's enough information to make a decision about its handling - showing an error that some feature is temporary unavailable and ask retry later or degrade the experience until the service has recovered.

Reply:
@SergeyMenshykh : that makes sense for this world, the original code was used in the context of SharePoint